main.go

```diff
@@ -317,8 +283,77 @@ func main() {
 	}
 }
 
-func before(c *cli.Context) (*state.HelmState, helmexec.Interface, error) {
-	file := c.GlobalString("file")
+func eachDesiredStateDo(c *cli.Context, converge func(*state.HelmState, helmexec.Interface) []error) error {
+	fileOrDirPath := c.GlobalString("file")
+	desiredStateFiles, err := findDesiredStateFiles(fileOrDirPath)
+	if err != nil {
+		return err
+	}
+	for _, f := range desiredStateFiles {
+		state, helm, err := loadDesiredStateFromFile(c, f)
+		if err != nil {
+			return err
+		}
+		errs := converge(state, helm)
+		if err := clean(state, errs); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func findDesiredStateFiles(fileOrDirPath string) ([]string, error) {
+	var dir string
+
+	if fileOrDirPath != "" {
+		if !fileExists(fileOrDirPath) {
+			return []string{}, fmt.Errorf("state file named %s is not found", fileOrDirPath)
+		}
+
+		if !directoryExists(fileOrDirPath) {
+			return []string{fileOrDirPath}, nil
+		}
+
+		dir = fileOrDirPath
+	}
+
+	if dir == "" && fileExists(DefaultHelmfileDirectory) && directoryExists(DefaultHelmfileDirectory) {
+		dir = DefaultHelmfileDirectory
+	}
+
+	if dir != "" {
+		files, err := filepath.Glob(filepath.Join(dir, "*.yaml"))
+		if err != nil {
+			return []string{}, err
+		}
+		sort.Slice(files, func(i, j int) bool {
+			return files[i] < files[j]
+		})
+		return files, nil
+	}
+
+	if fileExists(DefaultHelmfile) {
```

Comment:
Should we complain if we have both a helmfile dir and a helmfile?  Or should we load both?  Currently it would ignore the helmfile and use the helmdir.

Reply:
That's a sneaky kind of behavior, so whatever we do, I hope we make it explicit and/or noisy. Maybe a warning + using the helmdir?

Reply:
Typically with apps that have a `conf.d` structure, they also support a main config file. That main config file then will then have an option to point to a `conf.d` folder. For example `systemd`. Now, I'm *not* advocating adding this to this PR. As far as I'm concerned, executing one or the other (or even both) is fine by me.

Reply:
Thanks for the feedbacks! Let me go forward with the immediate exit for now. I want to make it very explicit.

Reply:
Addressed in the last commit